handler.go

```diff
@@ -23,170 +26,159 @@ import (
-	if resp.StatusCode == 401 && ca.Config.HubUser == "" {
-		return fmt.Errorf("Unable to authorise without credentials. Please set 'HUB USER' and 'HUB PASSWORD' from your Cloudradar account")
-	} else if resp.StatusCode == 401 && ca.Config.HubUser != "" {
-		return fmt.Errorf("Unable to authorise with provided credentials. Please correct 'HUB USER' and 'HUB PASSWORD' according to your Cloudradar account")
-	} else if resp.StatusCode < 200 || resp.StatusCode >= 400 {
-		return fmt.Errorf("Got bad response status: %d, %s. If you have a proxy or firewall it may be blocking the connection", resp.StatusCode, resp.Status)
-	}
+func newEmptyFieldError(name string) error {
+	err := fmt.Errorf("unexpected empty field %s", name)
+	return errors.Wrap(err, "the field must be filled with details of your Cloudradar account")
+}
 
-	return nil
+func newFieldError(name string, err error) error {
+	return errors.Wrapf(err, "%s field verification failed", name)
 }
 
-func (ca *Cagent) PostResultsToHub(result Result) error {
-	ca.initHubHTTPClient()
+func (ca *Cagent) PostResultToHub(ctx context.Context, result *Result) error {
+	ca.initHubClientOnce()
 
+	if len(ca.Config.HubURL) == 0 {
+		return newEmptyFieldError("hub_url")
+	} else if u, err := url.Parse(ca.Config.HubURL); err != nil {
+		return newFieldError("hub_url", err)
+	} else if u.Scheme != "http" && u.Scheme != "https" {
+		err := fmt.Errorf("wrong scheme '%s', URL must start with http:// or https://", u.Scheme)
+		return newFieldError("hub_url", err)
+	}
 	b, err := json.Marshal(result)
 	if err != nil {
+		err = errors.Wrap(err, "failed to serialize result")
 		return err
 	}
 
 	var req *http.Request
-
 	if ca.Config.HubGzip {
-		var buffer bytes.Buffer
-		zw := gzip.NewWriter(&buffer)
-		zw.Write(b)
-		zw.Close()
-		req, err = http.NewRequest("POST", ca.Config.HubURL, &buffer)
+		buf := new(bytes.Buffer)
+		gzipped := gzip.NewWriter(buf)
+		if _, err := gzipped.Write(b); err != nil {
+			err = errors.Wrap(err, "failed to write into gzipped buffer")
+			return err
+		}
+		if err := gzipped.Close(); err != nil {
+			err = errors.Wrap(err, "failed to finalize gzipped buffer")
+			return err
+		}
+		req, _ = http.NewRequest("POST", ca.Config.HubURL, buf)
 		req.Header.Set("Content-Encoding", "gzip")
 	} else {
-		req, err = http.NewRequest("POST", ca.Config.HubURL, bytes.NewBuffer(b))
+		req, _ = http.NewRequest("POST", ca.Config.HubURL, bytes.NewBuffer(b))
 	}
-	req.Header.Set("Content-Type", "application/json")
-
 	if err != nil {
 		return err
 	}
-
+	req.Header.Set("Content-Type", "application/json")
 	req.Header.Add("User-Agent", ca.userAgent())
-
-	if ca.Config.HubUser != "" {
+	if len(ca.Config.HubUser) > 0 {
 		req.SetBasicAuth(ca.Config.HubUser, ca.Config.HubPassword)
 	}
-
-	resp, err := ca.hubHTTPClient.Do(req)
-
-	if err != nil {
+	req = req.WithContext(ctx)
+	resp, err := ca.hubClient.Do(req)
+	if err = ca.checkClientError(resp, err, "hub_user", "hub_password"); err != nil {
 		return err
```

Comment:
Shouldn't we wrap the error?

Reply:
`checkClientError` is the wrapper for error from `hubClient.Do(req)`. It will parse it and response and format in a proper way. You suggest that a wrapper here could capture the stack and that could be useful later?

Reply:
@xlab Please don't simply resolve the conversation, especially when you ask another question in the.

> You suggest that a wrapper here could capture the stack and that could be useful later?

From my personal experience, it can be very helpful to _always_ wrap an error. In case of debugging, it can safe a lot of time to immediately have the full stack. Also it's an easy pattern to follow. If you return an error from a function call, wrap it.